plugins/scripting/scripting-ide-services/src/org/jetbrains/kotlin/scripting/ide_services/compiler/analyzer.kt

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.scripting.ide_services.compiler
+
+import org.jetbrains.kotlin.scripting.compiler.plugin.repl.DelegatePackageMemberDeclarationProvider
+import org.jetbrains.kotlin.cli.jvm.compiler.CliBindingTrace
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM
+import org.jetbrains.kotlin.container.ComponentProvider
+import org.jetbrains.kotlin.container.get
+import org.jetbrains.kotlin.container.getService
+import org.jetbrains.kotlin.descriptors.ClassDescriptorWithResolutionScopes
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.descriptors.impl.ModuleDescriptorImpl
+import org.jetbrains.kotlin.diagnostics.Severity
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.*
+import org.jetbrains.kotlin.resolve.calls.smartcasts.DataFlowInfoFactory
+import org.jetbrains.kotlin.resolve.diagnostics.Diagnostics
+import org.jetbrains.kotlin.resolve.lazy.*
+import org.jetbrains.kotlin.resolve.lazy.data.KtClassLikeInfo
+import org.jetbrains.kotlin.resolve.lazy.declarations.*
+import org.jetbrains.kotlin.resolve.scopes.ImportingScope
+import org.jetbrains.kotlin.resolve.scopes.utils.parentsWithSelf
+import org.jetbrains.kotlin.resolve.scopes.utils.replaceImportingScopes
+import org.jetbrains.kotlin.scripting.compiler.plugin.impl.KotlinResolutionFacade
+import org.jetbrains.kotlin.scripting.definitions.ScriptPriorities
+import org.jetbrains.kotlin.scripting.ide_services.util.ReplHistory
+import kotlin.script.experimental.api.SourceCode
+
+class CodeAnalyzer(private val environment: KotlinCoreEnvironment) {
+    private val container: ComponentProvider
+    private val topDownAnalysisContext: TopDownAnalysisContext
+    private val topDownAnalyzer: LazyTopDownAnalyzer
+    private val resolveSession: ResolveSession
+    private val scriptDeclarationFactory: ScriptMutableDeclarationProviderFactory
+    private val replState = ResettableAnalyzerState()
+
+    val module: ModuleDescriptorImpl
+
+    val trace: BindingTraceContext = CliBindingTrace()
```

Comment:
I'd still pass it as a parameter. This way it would be easier to reuse the Analyzer in the old code.

Reply:
I think that the reason of passing it as a parameter will become cleaner after reconsidering the project structure